bee-network/src/swarm/protocols/gossip/gossip_io.rs

```diff
@@ -0,0 +1,149 @@
+// Copyright 2020-2021 IOTA Stiftung
+// SPDX-License-Identifier: Apache-2.0
+
+use crate::{
+    alias,
+    service::event::{InternalEvent, InternalEventSender},
+};
+
+use futures::{
+    io::{ReadHalf, WriteHalf},
+    AsyncRead, AsyncReadExt, AsyncWrite, AsyncWriteExt, StreamExt,
+};
+use libp2p::{swarm::NegotiatedSubstream, PeerId};
+use log::*;
+use tokio::sync::mpsc;
+use tokio_stream::wrappers::UnboundedReceiverStream;
+
+const MSG_BUFFER_SIZE: usize = 32768;
+
+/// A shorthand for an unbounded channel sender.
+pub type GossipSender = mpsc::UnboundedSender<Vec<u8>>;
+
+/// A shorthand for an unbounded channel receiver.
+pub type GossipReceiver = UnboundedReceiverStream<Vec<u8>>;
+
+pub fn gossip_channel() -> (GossipSender, GossipReceiver) {
+    let (sender, receiver) = mpsc::unbounded_channel();
+    (sender, UnboundedReceiverStream::new(receiver))
+}
+
+pub fn spawn_gossip_in_processor(
+    peer_id: PeerId,
+    mut reader: ReadHalf<NegotiatedSubstream>,
+    incoming_gossip_sender: GossipSender,
+    internal_event_sender: InternalEventSender,
+) {
+    tokio::spawn(async move {
+        let mut msg_buf = vec![0u8; MSG_BUFFER_SIZE];
+        let mut msg_len = 0;
+
+        loop {
+            if recv_valid_message(&mut reader, &mut msg_buf, &mut msg_len).await {
+                if incoming_gossip_sender.send(msg_buf[..msg_len].to_vec()).is_err() {
+                    debug!("gossip-in: receiver dropped locally.");
+
+                    // The receiver of this channel was dropped, maybe due to a shutdown. There is nothing we can do to
+                    // salvage this situation, hence we drop the connection.
+                    break;
+                }
+            } else {
+                debug!("gossip-in: stream closed remotely.");
+
+                // NB: The network service will not shut down before it has received the `ProtocolDropped` event from
+                // all once connected peers, hence if the following send fails, then it must be
+                // considered a bug.
+
+                // The remote peer dropped the connection.
+                internal_event_sender
+                    .send(InternalEvent::ProtocolDropped { peer_id })
+                    .expect("The service must not shutdown as long as there are gossip tasks running.");
+
+                break;
+            }
+        }
+
+        // Reasons why this task might end:
+        // (1) The remote dropped the TCP connection.
+        // (2) The local dropped the gossip_in receiver channel.
+
+        debug!("gossip-in: exiting gossip-in processor for {}.", alias!(peer_id));
+    });
+}
+
+async fn recv_valid_message<S>(stream: &mut S, message: &mut [u8], message_len: &mut usize) -> bool
```

Comment:
Rather than having the function take `message_len: &mut usize`, it would be better to have it return `Option<usize>`. Otherwise, you're leaving room for additional, unspecified state to be represented (length change on read failure). At that point, however, I question the purpose of this function.

Reply:
Oh man, that method is already so old. Changed!